src/ucp/core/ucp_request.c

```diff
@@ -330,101 +330,37 @@ static void ucp_request_mem_invalidate_completion(uct_completion_t *comp)
-            ucs_assert(ucp_ep_get_iface_attr(ep, lane)->cap.flags &
-                       UCT_IFACE_FLAG_GET_ZCOPY);
-            ucs_assert(ucp_ep_md_attr(ep, lane)->cap.flags &
-                       UCT_MD_FLAG_INVALIDATE);
-            inv_map |= UCS_BIT(ucp_ep_md_index(ep, lane));
-        }
-    }
-
-    return inv_map & req->send.state.dt.dt.contig.md_map;
-}
-
 void ucp_request_dt_invalidate(ucp_request_t *req, ucs_status_t status)
 {
-    uct_md_mem_dereg_params_t params = {
-        .field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH |
-                      UCT_MD_MEM_DEREG_FIELD_FLAGS |
-                      UCT_MD_MEM_DEREG_FIELD_COMPLETION,
-        .flags      = UCT_MD_MEM_DEREG_FLAG_INVALIDATE,
-        .comp       = &req->send.state.uct_comp
-    };
     ucp_worker_h worker   = req->send.ep->worker;
     ucp_context_h context = worker->context;
-    uct_mem_h *uct_memh   = req->send.state.dt.dt.contig.memh;
-    ucp_md_map_t invalidate_map;
-    unsigned md_index;
-    unsigned memh_index;
 
     ucs_assert(status != UCS_OK);
     ucs_assert(ucp_ep_config(req->send.ep)->key.err_mode !=
                UCP_ERR_HANDLING_MODE_NONE);
     ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
 
-    invalidate_map                  = ucp_request_get_invalidation_map(req);
     req->send.ep                    = NULL;
-    req->send.state.uct_comp.count  = 1;
-    req->send.state.uct_comp.func   = ucp_request_mem_invalidate_completion;
-    req->send.state.uct_comp.status = UCS_OK;
     req->send.invalidate.worker     = worker;
     req->status                     = status;
 
-    ucp_trace_req(req, "mem dereg buffer md_map 0x%"PRIx64, invalidate_map);
-    /* dereg all lanes except for 'invalidate_map' */
-    ucp_mem_rereg_mds(context, invalidate_map, NULL, 0, 0, NULL,
-                      UCS_MEMORY_TYPE_HOST, NULL,
-                      req->send.state.dt.dt.contig.memh,
-                      &req->send.state.dt.dt.contig.md_map);
-    ucp_trace_req(req, "mem invalidate buffer md_map 0x%"PRIx64,
-                  req->send.state.dt.dt.contig.md_map);
-
-    memh_index = 0;
-    ucs_log_indent(1);
-    ucs_for_each_bit(md_index, req->send.state.dt.dt.contig.md_map) {
-        ucs_trace_req("invalidating memh[%d]=%p from md[%d]", memh_index,
-                      uct_memh[memh_index], md_index);
-        req->send.state.uct_comp.count++;
-        params.memh = uct_memh[memh_index];
-        status      = uct_md_mem_dereg_v2(context->tl_mds[md_index].md,
-                                          &params);
-        if (status != UCS_OK) {
-            ucs_warn("failed to dereg from md[%d]=%s: %s", md_index,
-                     context->tl_mds[md_index].rsc.md_name,
-                     ucs_status_string(status));
-            req->send.state.uct_comp.count--;
-        }
-        memh_index++;
+    if (req->send.state.dt.dt.contig.memh != NULL) {
+        ucp_memh_invalidate(context, req->send.state.dt.dt.contig.memh,
```

Comment:
where would we pass UCT_MD_MEM_DEREG_FLAG_INVALIDATE to the UCT-level deregistration?

Reply:
> Maybe we still have a chance to obsolete UCT_MD_MEM_DEREG_FLAG_INVALIDATE/UCT_MD_MEM_DEREG_FIELD_COMPLETION and leave only UCT_MD_MKEY_PACK_FLAG_INVALIDATE to control invalidation behavior? @evgeny-leksikov?

in this case we must limit UCT reg implemetnation to guarantee that every call returns unique key, i.e. transport cannot cache it.

Reply:
don't we want to remove transport caches and leave only UCP level cache?

Reply:
Evgeny probably meant "transport cannot recycle the remote key value"

Reply:
> don't we want to remove transport caches and leave only UCP level cache?

AFAIU we remove rcache from IB implementation only but in theory we cannot control all transports

> Evgeny probably meant "transport cannot recycle the remote key value"

I dont think we can avoid recycling all together, but removing existing flags progibites return the same rkey value for several "reg" calls in row similar like rcache behavior.

Reply:
since INVALIDATE is v2 API, we can change it to by synchronic, and remove completion for it altogether - but only after removing rcache from IB MD